src/protocol/ascii.rs

```diff
@@ -301,11 +359,23 @@ impl AsciiProtocol<Stream> {
         }
     }
 
-    pub(super) fn gets<V: FromMemcacheValueExt>(&mut self, keys: &[&str]) -> Result<HashMap<String, V>, MemcacheError> {
-        for key in keys {
-            check_key_len(key)?;
+    pub(super) fn gets<V: FromMemcacheValueExt, K: AsRef<str>>(
+        &mut self,
+        keys: &[K],
+    ) -> Result<HashMap<String, V>, MemcacheError> {
+        for k in keys.iter() {
+            check_key_len(k.as_ref())?;
+        }
+
+        let mut writer = BufWriter::new(self.reader.get_mut());
```

Comment:
Perhaps I wasn't clear, I meant that we should put the `BufWriter` in the `AsciiProtocol` struct. Use [try_clone](https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.try_clone) method on the streams and wrap them in `BufReader` and `BufWriter`.

```rust
struct AsciiProtocol<C> {
    reader: CappedLineReader<C>,
    writer: BufWriter<C>
}
```

Reply:
I tried something like that, but I discovered we can't use BufWriter for any code that writes values to the stream because the values are of the type ``ToMemcacheValue<Stream>``, which means the ``write_to(W)`` methods all expect W to be a ``Stream``, not a ``BufWriter``. At some point we might want to think about a less rigid design.

Reply:
That is because the impl is for `AsciiProtocol<Stream>`. If you make it `impl<C: Read + Write> AsciiProtocol<C>` and change all the ToMemcacheValue<Stream> to `ToMemcacheValue<C>`, I think you should be good to go.

Reply:
Would that work? The public interface (Client) is littered with ``ToMemcacheValue<Stream>``. Doesn't that mean we're stuck with using ``ToMemcacheValue<Stream>`` values unless we change the public interface? If we want to do that, I would prefer to start a new PR.

Reply:
FWIW, it just occurred to me that we could probably squeeze ``BufWriter`` (and maybe ``BufReader``?) inside the ``Stream`` enum, thus making everything use buffers without changing the public interface. That would be a new PR for sure though.

Reply:
We wouldn't have to change the public interface. Try this out.
```diff
diff --git a/src/protocol/ascii.rs b/src/protocol/ascii.rs
index 1b81ec4..c4b162e 100644
--- a/src/protocol/ascii.rs
+++ b/src/protocol/ascii.rs
@@ -6,7 +6,6 @@ use super::check_key_len;
 use client::Stats;
 use error::{ClientError, CommandError, MemcacheError, ServerError};
 use std::borrow::Cow;
-use stream::Stream;
 use value::{FromMemcacheValueExt, ToMemcacheValue};
 
 #[derive(Default)]
@@ -136,26 +135,50 @@ impl<C: Read> CappedLineReader<C> {
     }
 }
 
+pub struct MyBufWriter<C: Read + Write>(BufWriter<C>);
+
+impl<C: Read + Write> MyBufWriter<C> {
+    fn  get_mut(&mut self) -> &mut C {
+        self.0.get_mut()
+    }
+}
+
+impl<C: Read + Write> Read for MyBufWriter<C> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.0.get_mut().read(buf)
+    }
+}
+
+impl<C: Read + Write> Write for MyBufWriter<C> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.0.write(buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.0.flush()
+    }
+}
+
 pub struct AsciiProtocol<C: Read + Write + Sized> {
-    reader: CappedLineReader<C>,
+    stream: CappedLineReader<MyBufWriter<C>>,
 }
 
-impl AsciiProtocol<Stream> {
-    pub(crate) fn new(stream: Stream) -> Self {
+impl<C: Read + Write> AsciiProtocol<C> {
+    pub(crate) fn new(stream: C) -> Self {
         Self {
-            reader: CappedLineReader::new(stream),
+            stream: CappedLineReader::new(MyBufWriter(BufWriter::with_capacity(4096, stream))),
         }
     }
 
-    pub(crate) fn stream(&mut self) -> &mut Stream {
-        self.reader.get_mut()
+    pub(crate) fn stream(&mut self) -> &mut C {
+        self.stream.get_mut().get_mut()
     }
 
     pub(super) fn auth(&mut self, username: &str, password: &str) -> Result<(), MemcacheError> {
         return self.set("auth", format!("{} {}", username, password), 0);
     }
 
-    fn store<V: ToMemcacheValue<Stream>>(
+    fn store<V: ToMemcacheValue<C>>(
         &mut self,
         command: StoreCommand,
         key: &str,
@@ -165,7 +188,7 @@ impl AsciiProtocol<Stream> {
         Ok(self.stores(command, Some((key, value)), options)?)
     }
 
-    fn stores<V: ToMemcacheValue<Stream>, K: AsRef<str>, I: IntoIterator<Item = (K, V)>>(
+    fn stores<V: ToMemcacheValue<C>, K: AsRef<str>, I: IntoIterator<Item = (K, V)>>(
         &mut self,
         command: StoreCommand,
         entries: I,
@@ -187,7 +210,7 @@ impl AsciiProtocol<Stream> {
             check_key_len(key)?;
             if options.cas.is_some() {
                 write!(
-                    self.reader.get_mut(),
+                    self.stream.get_mut(),
                     "{command} {key} {flags} {exptime} {vlen} {cas}{noreply}\r\n",
                     command = command,
                     key = key,
@@ -199,7 +222,7 @@ impl AsciiProtocol<Stream> {
                 )?;
             } else {
                 write!(
-                    self.reader.get_mut(),
+                    self.stream.get_mut(),
                     "{command} {key} {flags} {exptime} {vlen}{noreply}\r\n",
                     command = command,
                     key = key,
@@ -210,13 +233,13 @@ impl AsciiProtocol<Stream> {
                 )?;
             }
 
-            value.write_to(self.reader.get_mut())?;
-            self.reader.get_mut().write_all(b"\r\n")?;
+            value.write_to(self.stream.get_mut().get_mut())?;
+            self.stream.get_mut().write_all(b"\r\n")?;
             sent_count += 1;
         }
 
         // Flush now that all the requests have been written.
-        self.reader.get_mut().flush()?;
+        self.stream.get_mut().flush()?;
 
         if options.noreply {
             return Ok(true);
@@ -230,7 +253,7 @@ impl AsciiProtocol<Stream> {
         let mut final_result = Ok(true);
 
         for _ in 0..sent_count {
-            let one_result = self.reader.read_line(|response| {
+            let one_result = self.stream.read_line(|response| {
                 let response = MemcacheError::try_from(response)?;
                 match response {
                     "STORED\r\n" => Ok(true),
@@ -261,9 +284,9 @@ impl AsciiProtocol<Stream> {
     }
 
     pub(super) fn version(&mut self) -> Result<String, MemcacheError> {
-        self.reader.get_mut().write_all(b"version\r\n")?;
-        self.reader.get_mut().flush()?;
-        self.reader.read_line(|response| {
+        self.stream.get_mut().write_all(b"version\r\n")?;
+        self.stream.get_mut().flush()?;
+        self.stream.read_line(|response| {
             let response = MemcacheError::try_from(response)?;
             if !response.starts_with("VERSION") {
                 Err(ServerError::BadResponse(Cow::Owned(response.into())))?
@@ -274,7 +297,7 @@ impl AsciiProtocol<Stream> {
     }
 
     fn parse_ok_response(&mut self) -> Result<(), MemcacheError> {
-        self.reader.read_line(|response| {
+        self.stream.read_line(|response| {
             let response = MemcacheError::try_from(response)?;
             if response == "OK\r\n" {
                 Ok(())
@@ -285,18 +308,18 @@ impl AsciiProtocol<Stream> {
     }
 
     pub(super) fn flush(&mut self) -> Result<(), MemcacheError> {
-        write!(self.reader.get_mut(), "flush_all\r\n")?;
+        write!(self.stream.get_mut(), "flush_all\r\n")?;
         self.parse_ok_response()
     }
 
     pub(super) fn flush_with_delay(&mut self, delay: u32) -> Result<(), MemcacheError> {
-        write!(self.reader.get_mut(), "flush_all {}\r\n", delay)?;
-        self.reader.get_mut().flush()?;
+        write!(self.stream.get_mut(), "flush_all {}\r\n", delay)?;
+        self.stream.get_mut().flush()?;
         self.parse_ok_response()
     }
 
     pub(super) fn get<V: FromMemcacheValueExt>(&mut self, key: &str) -> Result<Option<V>, MemcacheError> {
-        write!(self.reader.get_mut(), "get {}\r\n", key)?;
+        write!(self.stream.get_mut(), "get {}\r\n", key)?;
 
         if let Some((k, v)) = self.parse_get_response(false)? {
             if k != key {
@@ -317,7 +340,7 @@ impl AsciiProtocol<Stream> {
         &mut self,
         has_cas: bool,
     ) -> Result<Option<(String, V)>, MemcacheError> {
-        let result = self.reader.read_line(|buf| {
+        let result = self.stream.read_line(|buf| {
             let buf = MemcacheError::try_from(buf)?;
             if buf == END {
                 return Ok(None);
@@ -344,7 +367,7 @@ impl AsciiProtocol<Stream> {
         match result {
             Some((key, flags, length, cas)) => {
                 let mut value = vec![0u8; length + 2];
-                self.reader.read_exact(value.as_mut_slice())?;
+                self.stream.read_exact(value.as_mut_slice())?;
                 if &value[length..] != b"\r\n" {
                     return Err(ServerError::BadResponse(Cow::Owned(String::from_utf8(value)?)))?;
                 }
@@ -367,7 +390,7 @@ impl AsciiProtocol<Stream> {
             check_key_len(k.as_ref())?;
         }
 
-        let mut writer = BufWriter::new(self.reader.get_mut());
+        let mut writer = BufWriter::new(self.stream.get_mut());
         writer.write_all(b"gets")?;
         for k in keys.iter() {
             writer.write_all(b" ")?;
@@ -391,7 +414,7 @@ impl AsciiProtocol<Stream> {
         Err(ServerError::BadResponse(Cow::Borrowed("Expected end of gets response")))?
     }
 
-    pub(super) fn cas<V: ToMemcacheValue<Stream>>(
+    pub(super) fn cas<V: ToMemcacheValue<C>>(
         &mut self,
         key: &str,
         value: V,
@@ -412,7 +435,7 @@ impl AsciiProtocol<Stream> {
         }
     }
 
-    pub(super) fn set<V: ToMemcacheValue<Stream>>(
+    pub(super) fn set<V: ToMemcacheValue<C>>(
         &mut self,
         key: &str,
         value: V,
@@ -425,7 +448,7 @@ impl AsciiProtocol<Stream> {
         self.store(StoreCommand::Set, key, value, &options).map(|_| ())
     }
 
-    pub(super) fn sets<V: ToMemcacheValue<Stream>, K: AsRef<str>, I: IntoIterator<Item = (K, V)>>(
+    pub(super) fn sets<V: ToMemcacheValue<C>, K: AsRef<str>, I: IntoIterator<Item = (K, V)>>(
         &mut self,
         entries: I,
         expiration: u32,
@@ -437,7 +460,7 @@ impl AsciiProtocol<Stream> {
         self.stores(StoreCommand::Set, entries, &options).map(|_| ())
     }
 
-    pub(super) fn add<V: ToMemcacheValue<Stream>>(
+    pub(super) fn add<V: ToMemcacheValue<C>>(
         &mut self,
         key: &str,
         value: V,
@@ -450,7 +473,7 @@ impl AsciiProtocol<Stream> {
         self.store(StoreCommand::Add, key, value, &options).map(|_| ())
     }
 
-    pub(super) fn replace<V: ToMemcacheValue<Stream>>(
+    pub(super) fn replace<V: ToMemcacheValue<C>>(
         &mut self,
         key: &str,
         value: V,
@@ -463,13 +486,13 @@ impl AsciiProtocol<Stream> {
         self.store(StoreCommand::Replace, key, value, &options).map(|_| ())
     }
 
-    pub(super) fn append<V: ToMemcacheValue<Stream>>(&mut self, key: &str, value: V) -> Result<(), MemcacheError> {
+    pub(super) fn append<V: ToMemcacheValue<C>>(&mut self, key: &str, value: V) -> Result<(), MemcacheError> {
         check_key_len(key)?;
         self.store(StoreCommand::Append, key, value, &Default::default())
             .map(|_| ())
     }
 
-    pub(super) fn prepend<V: ToMemcacheValue<Stream>>(&mut self, key: &str, value: V) -> Result<(), MemcacheError> {
+    pub(super) fn prepend<V: ToMemcacheValue<C>>(&mut self, key: &str, value: V) -> Result<(), MemcacheError> {
         check_key_len(key)?;
         self.store(StoreCommand::Prepend, key, value, &Default::default())
             .map(|_| ())
@@ -482,10 +505,10 @@ impl AsciiProtocol<Stream> {
         }
 
         for k in keys {
-            write!(self.reader.get_mut(), "delete {}\r\n", k.as_ref())?;
+            write!(self.stream.get_mut(), "delete {}\r\n", k.as_ref())?;
         }
         // Flush now that all the requests have been written.
-        self.reader.get_mut().flush()?;
+        self.stream.get_mut().flush()?;
 
         // Receive all the responses. If there were errors, return the first.
 
@@ -493,7 +516,7 @@ impl AsciiProtocol<Stream> {
 
         for _ in 0..keys.len() {
             let one_result = self
-                .reader
+                .stream
                 .read_line(|response| match MemcacheError::try_from(response) {
                     Ok(s) => {
                         if s == "DELETED\r\n" {
@@ -530,7 +553,7 @@ impl AsciiProtocol<Stream> {
     }
 
     fn parse_u64_response(&mut self) -> Result<u64, MemcacheError> {
-        self.reader.read_line(|response| {
+        self.stream.read_line(|response| {
             let s = MemcacheError::try_from(response)?;
             Ok(s.trim_end_matches("\r\n").parse::<u64>()?)
         })
@@ -538,21 +561,21 @@ impl AsciiProtocol<Stream> {
 
     pub(super) fn increment(&mut self, key: &str, amount: u64) -> Result<u64, MemcacheError> {
         check_key_len(key)?;
-        write!(self.reader.get_mut(), "incr {} {}\r\n", key, amount)?;
+        write!(self.stream.get_mut(), "incr {} {}\r\n", key, amount)?;
         self.parse_u64_response()
     }
 
     pub(super) fn decrement(&mut self, key: &str, amount: u64) -> Result<u64, MemcacheError> {
         check_key_len(key)?;
-        write!(self.reader.get_mut(), "decr {} {}\r\n", key, amount)?;
+        write!(self.stream.get_mut(), "decr {} {}\r\n", key, amount)?;
         self.parse_u64_response()
     }
 
     pub(super) fn touch(&mut self, key: &str, expiration: u32) -> Result<bool, MemcacheError> {
         check_key_len(key)?;
-        write!(self.reader.get_mut(), "touch {} {}\r\n", key, expiration)?;
-        self.reader.get_mut().flush()?;
-        self.reader
+        write!(self.stream.get_mut(), "touch {} {}\r\n", key, expiration)?;
+        self.stream.get_mut().flush()?;
+        self.stream
             .read_line(|response| match MemcacheError::try_from(response) {
                 Ok(s) => {
                     if s == "TOUCHED\r\n" {
@@ -567,8 +590,8 @@ impl AsciiProtocol<Stream> {
     }
 
     pub(super) fn stats(&mut self) -> Result<Stats, MemcacheError> {
-        self.reader.get_mut().write_all(b"stats\r\n")?;
-        self.reader.get_mut().flush()?;
+        self.stream.get_mut().write_all(b"stats\r\n")?;
+        self.stream.get_mut().flush()?;
 
         enum Loop {
             Break,
@@ -577,7 +600,7 @@ impl AsciiProtocol<Stream> {
 
         let mut stats: Stats = HashMap::new();
         loop {
-            let status = self.reader.read_line(|response| {
+            let status = self.stream.read_line(|response| {
                 if response != END {
                     return Ok(Loop::Break);
                 }
```

Reply:
> Doesn't that mean we're stuck with using ToMemcacheValue<Stream> values unless we change the public interface?

No. AsciiProtocol works for all C where Read + Write, just that it's implemented for `Stream`. We will generalize it. I was not sure whether this would work, so tried it out myself. The following seems to work. 

```diff
diff --git a/src/protocol/ascii.rs b/src/protocol/ascii.rs
index 1b81ec4..be28580 100644
--- a/src/protocol/ascii.rs
+++ b/src/protocol/ascii.rs
@@ -6,7 +6,6 @@ use super::check_key_len;
 use client::Stats;
 use error::{ClientError, CommandError, MemcacheError, ServerError};
 use std::borrow::Cow;
-use stream::Stream;
 use value::{FromMemcacheValueExt, ToMemcacheValue};
 
 #[derive(Default)]
@@ -136,26 +135,50 @@ impl<C: Read> CappedLineReader<C> {
     }
 }
 
+pub struct MyBufWriter<C: Read + Write>(BufWriter<C>);
+
+impl<C: Read + Write> MyBufWriter<C> {
+    fn  get_mut(&mut self) -> &mut C {
+        self.0.get_mut()
+    }
+}
+
+impl<C: Read + Write> Read for MyBufWriter<C> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.0.get_mut().read(buf)
+    }
+}
+
+impl<C: Read + Write> Write for MyBufWriter<C> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.0.write(buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.0.flush()
+    }
+}
+
 pub struct AsciiProtocol<C: Read + Write + Sized> {
-    reader: CappedLineReader<C>,
+    stream: CappedLineReader<MyBufWriter<C>>,
 }
 
-impl AsciiProtocol<Stream> {
-    pub(crate) fn new(stream: Stream) -> Self {
+impl<C: Read + Write> AsciiProtocol<C> {
+    pub(crate) fn new(stream: C) -> Self {
         Self {
-            reader: CappedLineReader::new(stream),
+            stream: CappedLineReader::new(MyBufWriter(BufWriter::with_capacity(4096, stream))),
         }
     }
 
-    pub(crate) fn stream(&mut self) -> &mut Stream {
-        self.reader.get_mut()
+    pub(crate) fn stream(&mut self) -> &mut C {
+        self.stream.get_mut().get_mut()
     }
 
     pub(super) fn auth(&mut self, username: &str, password: &str) -> Result<(), MemcacheError> {
         return self.set("auth", format!("{} {}", username, password), 0);
     }
 
-    fn store<V: ToMemcacheValue<Stream>>(
+    fn store<V: ToMemcacheValue<C>>(
         &mut self,
         command: StoreCommand,
         key: &str,
@@ -165,7 +188,7 @@ impl AsciiProtocol<Stream> {
         Ok(self.stores(command, Some((key, value)), options)?)
     }
 
-    fn stores<V: ToMemcacheValue<Stream>, K: AsRef<str>, I: IntoIterator<Item = (K, V)>>(
+    fn stores<V: ToMemcacheValue<C>, K: AsRef<str>, I: IntoIterator<Item = (K, V)>>(
         &mut self,
         command: StoreCommand,
         entries: I,
@@ -187,7 +210,7 @@ impl AsciiProtocol<Stream> {
             check_key_len(key)?;
             if options.cas.is_some() {
                 write!(
-                    self.reader.get_mut(),
+                    self.stream.get_mut(),
                     "{command} {key} {flags} {exptime} {vlen} {cas}{noreply}\r\n",
                     command = command,
                     key = key,
@@ -199,7 +222,7 @@ impl AsciiProtocol<Stream> {
                 )?;
             } else {
                 write!(
-                    self.reader.get_mut(),
+                    self.stream.get_mut(),
                     "{command} {key} {flags} {exptime} {vlen}{noreply}\r\n",
                     command = command,
                     key = key,
@@ -210,13 +233,13 @@ impl AsciiProtocol<Stream> {
                 )?;
             }
 
-            value.write_to(self.reader.get_mut())?;
-            self.reader.get_mut().write_all(b"\r\n")?;
+            value.write_to(self.stream.get_mut().get_mut())?;
+            self.stream.get_mut().write_all(b"\r\n")?;
             sent_count += 1;
         }
 
         // Flush now that all the requests have been written.
-        self.reader.get_mut().flush()?;
+        self.stream.get_mut().flush()?;
 
         if options.noreply {
             return Ok(true);
@@ -230,7 +253,7 @@ impl AsciiProtocol<Stream> {
         let mut final_result = Ok(true);
 
         for _ in 0..sent_count {
-            let one_result = self.reader.read_line(|response| {
+            let one_result = self.stream.read_line(|response| {
                 let response = MemcacheError::try_from(response)?;
                 match response {
                     "STORED\r\n" => Ok(true),
@@ -261,9 +284,9 @@ impl AsciiProtocol<Stream> {
     }
 
     pub(super) fn version(&mut self) -> Result<String, MemcacheError> {
-        self.reader.get_mut().write_all(b"version\r\n")?;
-        self.reader.get_mut().flush()?;
-        self.reader.read_line(|response| {
+        self.stream.get_mut().write_all(b"version\r\n")?;
+        self.stream.get_mut().flush()?;
+        self.stream.read_line(|response| {
             let response = MemcacheError::try_from(response)?;
             if !response.starts_with("VERSION") {
                 Err(ServerError::BadResponse(Cow::Owned(response.into())))?
@@ -274,7 +297,7 @@ impl AsciiProtocol<Stream> {
     }
 
     fn parse_ok_response(&mut self) -> Result<(), MemcacheError> {
-        self.reader.read_line(|response| {
+        self.stream.read_line(|response| {
             let response = MemcacheError::try_from(response)?;
             if response == "OK\r\n" {
                 Ok(())
@@ -285,18 +308,18 @@ impl AsciiProtocol<Stream> {
     }
 
     pub(super) fn flush(&mut self) -> Result<(), MemcacheError> {
-        write!(self.reader.get_mut(), "flush_all\r\n")?;
+        write!(self.stream.get_mut(), "flush_all\r\n")?;
         self.parse_ok_response()
     }
 
     pub(super) fn flush_with_delay(&mut self, delay: u32) -> Result<(), MemcacheError> {
-        write!(self.reader.get_mut(), "flush_all {}\r\n", delay)?;
-        self.reader.get_mut().flush()?;
+        write!(self.stream.get_mut(), "flush_all {}\r\n", delay)?;
+        self.stream.get_mut().flush()?;
         self.parse_ok_response()
     }
 
     pub(super) fn get<V: FromMemcacheValueExt>(&mut self, key: &str) -> Result<Option<V>, MemcacheError> {
-        write!(self.reader.get_mut(), "get {}\r\n", key)?;
+        write!(self.stream.get_mut(), "get {}\r\n", key)?;
 
         if let Some((k, v)) = self.parse_get_response(false)? {
             if k != key {
@@ -317,7 +340,7 @@ impl AsciiProtocol<Stream> {
         &mut self,
         has_cas: bool,
     ) -> Result<Option<(String, V)>, MemcacheError> {
-        let result = self.reader.read_line(|buf| {
+        let result = self.stream.read_line(|buf| {
             let buf = MemcacheError::try_from(buf)?;
             if buf == END {
                 return Ok(None);
@@ -344,7 +367,7 @@ impl AsciiProtocol<Stream> {
         match result {
             Some((key, flags, length, cas)) => {
                 let mut value = vec![0u8; length + 2];
-                self.reader.read_exact(value.as_mut_slice())?;
+                self.stream.read_exact(value.as_mut_slice())?;
                 if &value[length..] != b"\r\n" {
                     return Err(ServerError::BadResponse(Cow::Owned(String::from_utf8(value)?)))?;
                 }
@@ -367,7 +390,7 @@ impl AsciiProtocol<Stream> {
             check_key_len(k.as_ref())?;
         }
 
-        let mut writer = BufWriter::new(self.reader.get_mut());
+        let writer = self.stream.get_mut();
         writer.write_all(b"gets")?;
         for k in keys.iter() {
             writer.write_all(b" ")?;
@@ -375,7 +398,6 @@ impl AsciiProtocol<Stream> {
         }
         writer.write_all(b"\r\n")?;
         writer.flush()?;
-        drop(writer);
 
         let mut result: HashMap<String, V> = HashMap::with_capacity(keys.len());
         // there will be atmost keys.len() "VALUE <...>" responses and one END response
@@ -391,7 +413,7 @@ impl AsciiProtocol<Stream> {
         Err(ServerError::BadResponse(Cow::Borrowed("Expected end of gets response")))?
     }
 
-    pub(super) fn cas<V: ToMemcacheValue<Stream>>(
+    pub(super) fn cas<V: ToMemcacheValue<C>>(
         &mut self,
         key: &str,
         value: V,
@@ -412,7 +434,7 @@ impl AsciiProtocol<Stream> {
         }
     }
 
-    pub(super) fn set<V: ToMemcacheValue<Stream>>(
+    pub(super) fn set<V: ToMemcacheValue<C>>(
         &mut self,
         key: &str,
         value: V,
@@ -425,7 +447,7 @@ impl AsciiProtocol<Stream> {
         self.store(StoreCommand::Set, key, value, &options).map(|_| ())
     }
 
-    pub(super) fn sets<V: ToMemcacheValue<Stream>, K: AsRef<str>, I: IntoIterator<Item = (K, V)>>(
+    pub(super) fn sets<V: ToMemcacheValue<C>, K: AsRef<str>, I: IntoIterator<Item = (K, V)>>(
         &mut self,
         entries: I,
         expiration: u32,
@@ -437,7 +459,7 @@ impl AsciiProtocol<Stream> {
         self.stores(StoreCommand::Set, entries, &options).map(|_| ())
     }
 
-    pub(super) fn add<V: ToMemcacheValue<Stream>>(
+    pub(super) fn add<V: ToMemcacheValue<C>>(
         &mut self,
         key: &str,
         value: V,
@@ -450,7 +472,7 @@ impl AsciiProtocol<Stream> {
         self.store(StoreCommand::Add, key, value, &options).map(|_| ())
     }
 
-    pub(super) fn replace<V: ToMemcacheValue<Stream>>(
+    pub(super) fn replace<V: ToMemcacheValue<C>>(
         &mut self,
         key: &str,
         value: V,
@@ -463,13 +485,13 @@ impl AsciiProtocol<Stream> {
         self.store(StoreCommand::Replace, key, value, &options).map(|_| ())
     }
 
-    pub(super) fn append<V: ToMemcacheValue<Stream>>(&mut self, key: &str, value: V) -> Result<(), MemcacheError> {
+    pub(super) fn append<V: ToMemcacheValue<C>>(&mut self, key: &str, value: V) -> Result<(), MemcacheError> {
         check_key_len(key)?;
         self.store(StoreCommand::Append, key, value, &Default::default())
             .map(|_| ())
     }
 
-    pub(super) fn prepend<V: ToMemcacheValue<Stream>>(&mut self, key: &str, value: V) -> Result<(), MemcacheError> {
+    pub(super) fn prepend<V: ToMemcacheValue<C>>(&mut self, key: &str, value: V) -> Result<(), MemcacheError> {
         check_key_len(key)?;
         self.store(StoreCommand::Prepend, key, value, &Default::default())
             .map(|_| ())
@@ -482,10 +504,10 @@ impl AsciiProtocol<Stream> {
         }
 
         for k in keys {
-            write!(self.reader.get_mut(), "delete {}\r\n", k.as_ref())?;
+            write!(self.stream.get_mut(), "delete {}\r\n", k.as_ref())?;
         }
         // Flush now that all the requests have been written.
-        self.reader.get_mut().flush()?;
+        self.stream.get_mut().flush()?;
 
         // Receive all the responses. If there were errors, return the first.
 
@@ -493,7 +515,7 @@ impl AsciiProtocol<Stream> {
 
         for _ in 0..keys.len() {
             let one_result = self
-                .reader
+                .stream
                 .read_line(|response| match MemcacheError::try_from(response) {
                     Ok(s) => {
                         if s == "DELETED\r\n" {
@@ -530,7 +552,7 @@ impl AsciiProtocol<Stream> {
     }
 
     fn parse_u64_response(&mut self) -> Result<u64, MemcacheError> {
-        self.reader.read_line(|response| {
+        self.stream.read_line(|response| {
             let s = MemcacheError::try_from(response)?;
             Ok(s.trim_end_matches("\r\n").parse::<u64>()?)
         })
@@ -538,21 +560,21 @@ impl AsciiProtocol<Stream> {
 
     pub(super) fn increment(&mut self, key: &str, amount: u64) -> Result<u64, MemcacheError> {
         check_key_len(key)?;
-        write!(self.reader.get_mut(), "incr {} {}\r\n", key, amount)?;
+        write!(self.stream.get_mut(), "incr {} {}\r\n", key, amount)?;
         self.parse_u64_response()
     }
 
     pub(super) fn decrement(&mut self, key: &str, amount: u64) -> Result<u64, MemcacheError> {
         check_key_len(key)?;
-        write!(self.reader.get_mut(), "decr {} {}\r\n", key, amount)?;
+        write!(self.stream.get_mut(), "decr {} {}\r\n", key, amount)?;
         self.parse_u64_response()
     }
 
     pub(super) fn touch(&mut self, key: &str, expiration: u32) -> Result<bool, MemcacheError> {
         check_key_len(key)?;
-        write!(self.reader.get_mut(), "touch {} {}\r\n", key, expiration)?;
-        self.reader.get_mut().flush()?;
-        self.reader
+        write!(self.stream.get_mut(), "touch {} {}\r\n", key, expiration)?;
+        self.stream.get_mut().flush()?;
+        self.stream
             .read_line(|response| match MemcacheError::try_from(response) {
                 Ok(s) => {
                     if s == "TOUCHED\r\n" {
@@ -567,8 +589,8 @@ impl AsciiProtocol<Stream> {
     }
 
     pub(super) fn stats(&mut self) -> Result<Stats, MemcacheError> {
-        self.reader.get_mut().write_all(b"stats\r\n")?;
-        self.reader.get_mut().flush()?;
+        self.stream.get_mut().write_all(b"stats\r\n")?;
+        self.stream.get_mut().flush()?;
 
         enum Loop {
             Break,
@@ -577,7 +599,7 @@ impl AsciiProtocol<Stream> {
 
         let mut stats: Stats = HashMap::new();
         loop {
-            let status = self.reader.read_line(|response| {
+            let status = self.stream.read_line(|response| {
                 if response != END {
                     return Ok(Loop::Break);
                 }
```

Reply:
Found the test will hang on flush command after apply this patch, will dig it latter.